src/lib/repositories/impl_oauth2/oauth2_repository_impl.py

```diff
@@ -0,0 +1,231 @@
+import bcrypt
+from injector import inject
+from sqlalchemy import text
+from sqlalchemy.engine.base import Engine
+import jwt
+from src.env_config import get_env_config_instance
+from src.utils.oauth2_util import (
+    build_client_credentials_access_token,
+    build_client_credentials_refresh_token,
+    build_user_credential_access_token,
+    build_user_credential_refresh_token,
+)
+from src.constants.oauth2 import GranType
+from src.utils.sql_oath2_queries import (
+    SQL_QUERY_TO_GET_USER_BY_USERNAME,
+    SQL_QUERY_TO_GET_CLIENT_BY_CLIENT_ID,
+    SQL_QUERY_TO_GET_CLIENT_BY_ID,
+    SQL_QUERY_TO_ADD_ACCESS_TOKEN,
+    SQL_QUERY_TO_ADD_REFRESH_TOKEN,
+    SQL_QUERY_TO_GET_REFRESH_TOKEN_BY_TOKEN,
+    SQL_QUERY_TO_GET_SCOPE_BY_CLIENT_ID,
+    SQL_QUERY_TO_DELETE_ACCESS_TOKEN_BY_REFRESH_TOKEN_ID,
+    SQL_QUERY_TO_DELETE_REFRESH_TOKEN,
+    SQL_QUERY_TO_GET_CLIENT_USER_BY_USERNAME_AND_CLIENT_ID,
+)
+
+
+class Oauth2Repository:
+    @inject
+    def __init__(self, engine: Engine):
+        self.engine = engine
+        self.env_config = get_env_config_instance()
+
+    def login_client(self, client_id, client_secret):
+        return self._login(
+            client_id=client_id,
+            client_secret=client_secret,
+            grant_type=GranType.CLIENT_CREDENTIALS,
+        )
+
+    def login_user(self, client_id, client_secret, username, password):
+        return self._login(
+            username=username,
+            password=password,
+            client_id=client_id,
+            client_secret=client_secret,
+            grant_type=GranType.PASSWORD,
+        )
+
+    def _login(
+        self,
+        client_id=None,
+        client_secret=None,
+        username=None,
+        password=None,
+        grant_type=None,
+    ):
+        self._validate_client_credentials(client_id, client_secret)
+
+        client = self._get_client_by_client_id(client_id)
+
+        if grant_type == GranType.CLIENT_CREDENTIALS:
+
+            client_scope = self._get_scope_by_app_client_id(client.id)
```

Comment:
This can be outside of the `if block`. I thought clients could have more than one scope?

Reply:
done.